src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/FunctionProbes/FunctionProbesStub.cs

```diff
@@ -10,8 +11,7 @@ public static class FunctionProbesStub
         private delegate void EnterProbeDelegate(ulong uniquifier, object[] args);
         private static readonly EnterProbeDelegate s_fixedEnterProbeDelegate = EnterProbeStub;
 
-        [ThreadStatic]
-        private static bool s_inProbe;
+        private static readonly AsyncLocal<int> s_inProbeCount = new();
```

Comment:
This need to remain a `ThreadStatic` otherwise we risk recursing probes. `*Local` types in .NET (e.g. `ThreadLocal` and `AsyncLocal`) operate via managed code outside of our control. That managed code may call methods which we have probes installed into, so the recursing check would cause recursion.

Reply:
> Updated FunctionProbes to support pausing capturing across async operations and added the NoProbeContext class to easily define a scope in which capturing is disabled.

Why is this change needed? Is it because event source writes out events using tasks?

Reply:
I noticed we were tracking String.Concat and other in `FunctionProbesManager.StartCapturingAsync`

Reply:
I reverted back to `ThreadStatic`, but I kept the ref count. We discussed offline for having a collection of excluded threads, but I think the ref count approach is simpler and doesn't involve any locking.